Wikipedia/Code/WMFRecentPagesDataSource.m

```diff
@@ -59,6 +59,26 @@ - (nonnull instancetype)initWithRecentPagesList:(MWKHistoryList*)recentPages sav
             [dataSource removeItemAtIndexPath:indexPath];
             [[NSNotificationCenter defaultCenter] addObserver:dataSource selector:@selector(rebuildSections) name:MWKHistoryListDidUpdateNotification object:recentPages];
         };
+
+        [self.KVOController observe:self.recentPages keyPath:WMF_SAFE_KEYPATH(self.recentPages, entries) options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionPrior block:^(WMFRecentPagesDataSource* observer, MWKHistoryList* object, NSDictionary* change) {
+            BOOL isPrior = [change[NSKeyValueChangeNotificationIsPriorKey] boolValue];
+            NSKeyValueChange changeKind = [change[NSKeyValueChangeKindKey] unsignedIntegerValue];
+            NSIndexSet* indexes = change[NSKeyValueChangeIndexesKey];
+
+            if (isPrior) {
```

Comment:
Yeah - I can't really. Since "items" isn't a real property in the data source, there wasn't a way to support observations. Of course this all goes away with Core Data (and removal of the lists classes), so I didn't like it - but wasn't too concerned.